### PR TITLE
Support Postgres with SSL certificates

### PR DESCRIPTION
**Pull request by bdunne**
part of #652

### CODE REVIEW

File: manageiq-operator/README.md [hunk 1]
```diff
@@ -110,7 +110,17 @@ $ podman push docker.io/example/manageiq-bundle:0.0.1
...
+```
+$ oc create secret generic postgresql-secrets \
+  --from-literal=dbname=vmdb_production \
+  --from-literal=hostname=YOUR_HOSTNAME \
+  --from-literal=port=5432 \
+  --from-literal=password=YOUR_PASSWORD_HERE \
+  --from-literal=username=YOUR_USERNAME_HERE \
+  --from-literal=sslmode=verify-full \ # optional
+  --from-file=rootcertificate=path/to/optional/certificate.pem # optional
+  ```
```
Comment (Fryguy) [inline manageiq-operator/README.md:123]: Minor, but the triple backticks are not aligned.

File: manageiq-operator/pkg/helpers/miq-components/orchestrator.go [hunk 2]
```diff
@@ -159,6 +160,35 @@ func addMessagingEnv(cr *miqv1alpha1.ManageIQ, c *corev1.Container) {
...
+	d.Spec.Template.Spec.Containers[0].Env = append(d.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{Name: "DATABASE_USER", ValueFrom: &corev1.EnvVarSource{SecretKeyRef: &corev1.SecretKeySelector{LocalObjectReference: corev1.LocalObjectReference{Name: postgresqlSecretName(cr)}, Key: "username"}}})
+
+	pgSecret := postgresqlSecret(cr, client)
+	if pgSecret.Data["sslmode"] != nil {
+		d.Spec.Template.Spec.Containers[0].Env = append(d.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{Name: "DATABASE_SSL_MODE", ValueFrom: &corev1.EnvVarSource{SecretKeyRef: &corev1.SecretKeySelector{LocalObjectReference: corev1.LocalObjectReference{Name: postgresqlSecretName(cr)}, Key: "sslmode"}}})
+	}
+	if pgSecret.Data["rootcertificate"] != nil {
+		d.Spec.Template.Spec.Containers[0].VolumeMounts = []corev1.VolumeMount{corev1.VolumeMount{Name: "miq-pg-root-certificate", MountPath: "/.postgresql", ReadOnly: true}}
```
Comment (Fryguy) [inline manageiq-operator/pkg/helpers/miq-components/orchestrator.go:177]: Does the `miq-` bit have to be in the name? (or is that the volume name from within the container)?

File: manageiq-operator/pkg/helpers/miq-components/orchestrator.go [hunk 3]
```diff
@@ -159,6 +160,35 @@ func addMessagingEnv(cr *miqv1alpha1.ManageIQ, c *corev1.Container) {
...
+	}
+}
+
+func postgresqlSecret(cr *miqv1alpha1.ManageIQ, client client.Client) *corev1.Secret {
+	secretKey := types.NamespacedName{Namespace: cr.Namespace, Name: postgresqlSecretName(cr)}
+	secret := &corev1.Secret{}
+	client.Get(context.TODO(), secretKey, secret)
+
+	return secret
+}
```
Comment (Fryguy) [inline manageiq-operator/pkg/helpers/miq-components/orchestrator.go:190]: Should this be in the other file where we define all of the secrets?

**Comment by Fryguy** [comment]
Overall LGTM...just some minor things

File: images/manageiq-base/container-assets/container_env [hunk 4]
```diff
@@ -9,4 +9,8 @@ function urlescape() {
 safeuser=$(urlescape ${DATABASE_USER})
 safepass=$(urlescape ${DATABASE_PASSWORD})
 
-export DATABASE_URL="postgresql://${safeuser}:${safepass}@${DATABASE_HOSTNAME}:${DATABASE_PORT}/${DATABASE_NAME}?encoding=utf8&pool=5&wait_timeout=5"
+ssl_mode=${DATABASE_SSL_MODE:-prefer}
+database_url="postgresql://${safeuser}:${safepass}@${DATABASE_HOSTNAME}:${DATABASE_PORT}/${DATABASE_NAME}?encoding=utf8&pool=5&wait_timeout=5&sslmode=${ssl_mode}"
+[[ -f /.postgresql/postgresql.crt ]] && database_url="$database_url&sslrootcert=/.postgresql/postgresql.crt"
```
Comment (NickLaMuro) [inline images/manageiq-base/container-assets/container_env:14]: To avoid extra "`if`" statements in this file (which I find hard to parse\*), does it make sense try using what was already available for environment variables from postgres?

> The location of the root certificate file and the CRL can be changed by setting the connection parameters `sslrootcert` and `sslcrl` or the environment variables `PGSSLROOTCERT` and `PGSSLCRL`.

(info pull from here:  https://www.postgresql.org/docs/10/libpq-ssl.html#LIBQ-SSL-CERTIFICATES )

I think there is a case to be made that it makes it more clear to do all variables in one fashion, so this isn't a hard stance I am making.  That said, it seems like instead of adding more logic here, it makes more sense to just add another environment variable to the process if `"rootcertificate"` is detected in the `orchestrator.go ` file.

* * *

If we do go this route, we could add a comment here mentioning that root certs are defined via the `orchestrator.go`, and that will set `PGSSLROOTCERT` for you.  Less places the path for the cert is hard coded as well (would only be in the `orchestrator` that way.

**Review by NickLaMuro — COMMENTED** [review]
I think this is fine based on what little I know about the `orchestrator.go` code base.

I really can't comment on elongated lines like this:

```diff
+ d.Spec.Template.Spec.Containers[0].Env = append(d.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{Name: "DATABASE_HOSTNAME", ValueFrom: &corev1.EnvVarSource{SecretKeyRef: &corev1.SecretKeySelector{LocalObjectReference: corev1.LocalObjectReference{Name: postgresqlSecretName(cr)}, Key: "hostname"}}})
```

Since I have next to zero knowledge about the libraries and facilities they are calling into, but seems it is mostly like it is just some refactoring, and a bit of IO tweaking for how we are setting configuring the orchestrator (correct me if I am wrong)

File: images/manageiq-base/container-assets/container_env [hunk 5]
```diff
@@ -9,4 +9,8 @@ function urlescape() {
 safeuser=$(urlescape ${DATABASE_USER})
 safepass=$(urlescape ${DATABASE_PASSWORD})
 
-export DATABASE_URL="postgresql://${safeuser}:${safepass}@${DATABASE_HOSTNAME}:${DATABASE_PORT}/${DATABASE_NAME}?encoding=utf8&pool=5&wait_timeout=5"
+ssl_mode=${DATABASE_SSL_MODE:-prefer}
+database_url="postgresql://${safeuser}:${safepass}@${DATABASE_HOSTNAME}:${DATABASE_PORT}/${DATABASE_NAME}?encoding=utf8&pool=5&wait_timeout=5&sslmode=${ssl_mode}"
```
[the remark this answers is not in the document]
Comment (NickLaMuro) [reply images/manageiq-base/container-assets/container_env:13]: At a minimum, we probably should move this up to the be next to the `safeuser` and `safepass` variable declarations.

File: images/manageiq-base/container-assets/container_env (hunk 4 above)
Comment (bdunne) [reply images/manageiq-base/container-assets/container_env:14]: I'm guessing we don't need to set `PGSSLROOTCERT` since I'm dropping it in the default location.

I figured that if this were on an appliance we would set this in `database.yml` anyway, so I put it in `DATABASE_URL` for consistency.

File: manageiq-operator/pkg/helpers/miq-components/orchestrator.go (hunk 3 above)
Comment (bdunne) [reply manageiq-operator/pkg/helpers/miq-components/orchestrator.go]: Moved

File: manageiq-operator/pkg/helpers/miq-components/orchestrator.go (hunk 2 above)
Comment (bdunne) [reply manageiq-operator/pkg/helpers/miq-components/orchestrator.go]: Done.

File: manageiq-operator/README.md (hunk 1 above)
Comment (bdunne) [reply manageiq-operator/README.md]: Fixed

File: images/manageiq-base/container-assets/container_env (hunk 4 above)
Comment (NickLaMuro) [reply images/manageiq-base/container-assets/container_env:13]: > I figured that if this were on an appliance we would set this in `database.yml` anyway, so I put it in `DATABASE_URL` for consistency.

This rational makes sense, and I am cool with that.  I wasn't hard set on my position here anyway, so I am fine with this staying as it is.

**Review by NickLaMuro — APPROVED** [review]
My concerns have been addressed, and everything else is probably out of my wheelhouse to be able to comment on properly.  :shipit:
